website_canonical_url/models/website.py

```diff
@@ -1,35 +1,63 @@
 # -*- coding: utf-8 -*-
+# Copyright 2018 Simone Orsi <simone.orsi@camptocamp.com>
 # Copyright initOS GmbH 2016
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from odoo import models, api
+from odoo import models, api, fields, exceptions, _
 from odoo.http import request
-from urlparse import urlparse, urlunparse
+from urlparse import urlparse, urljoin
 
 
 class Website(models.Model):
     _inherit = 'website'
 
+    canonical_domain = fields.Char(
+        help='Canonical domain is used to build unique canonical URLs '
+             'to make SEO happy.'
+    )
+
+    @api.constrains('canonical_domain')
+    def _check_canonical_domain(self):
+        domain = self.canonical_domain
+        if domain and not domain.startswith('http'):
+            raise exceptions.ValidationError(_(
+                'Canonical domain must contain protocol `http(s)://`'
+            ))
+
     @api.multi
     def get_canonical_url(self, req=None):
+        return urljoin(
+            self._get_canonical_domain(),
+            self._get_canonical_relative_url(req=req)
+        )
+
+    @api.multi
+    def _get_canonical_domain(self):
+        self.ensure_one()
+        if self.canonical_domain:
+            return self.canonical_domain
+        params = self.env['ir.config_parameter'].sudo()
+        return params.get_param('web.base.url')
+
+    @api.multi
+    def _get_canonical_relative_url(self, req=None):
         canonical_url = None
-        if req is None:  # pragma: no cover
-            req = request
-        if req and req.httprequest.path:
-            lang = self.env.lang
-            if lang == request.website.default_lang_code:
-                canonical_url = req.httprequest.path
-            else:
-                url_parts = urlparse(req.httprequest.path)
-                url_parts = list(url_parts)
-                # change the path of the url
-                url_parts[2] = lang + url_parts[2]
-                canonical_url = urlunparse(url_parts)
+        req = req or request
+        lang = getattr(req, 'lang', None) or self.env.lang
+        parsed = urlparse(req.httprequest.path)
+        canonical_url = parsed.path
+        if lang != req.website.default_lang_code:
+            for avail_lang, __ in self.get_languages():
+                avail_lang = avail_lang.encode('utf-8')
+                if canonical_url.startswith('/' + avail_lang):
+                    canonical_url = canonical_url[len(avail_lang) + 1:]
```

Comment:
I think the logic here is wrong.

If you're on the `it_IT` version of the website, the canonical URL for `http://example.com/it_IT/events?show_all=true` should be `http://example.com/it_IT/events`.

If we canonicalize only the default language version, search engines would never index other languages. The canonical tag **must** include the lang if it's not the default.

For you to understand better, the metatag used to indicate a page translations is [hreflang](https://support.google.com/webmasters/answer/189077), and Odoo ships it by default.

Reply:
ah, also the old implementation (like current) seems broken in the special case handling for `/page/homepage` check since is not considering the language and set the url always to `/`.